### ISSUE TITLE
RFC 3: using a dependency solver

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide-lock#10]
Here the idea is to use some standard Python dependency solver for creating the  `pyodide-lock.json` from multiple sources (Pyodide distribution, PyPI, third party sources etc).

For instance it could look as follows,

1. Expose a Simple JSON API for the Pyodide distribution (so that Python package managers could understand it). This should already be done, but I can't find where (cc @ryanking13 )
2. Define a list of indexes with a priority . Same idea as [`--extra-index-url` in pip](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-extra-index-url)
3. Run a standard Pythopn dependency solver on a list of input requirements, specifying `emscripten*_wasm32` as the platform (or whatever it is now). Here the best tool remains to be determined,
    - using solvers from pip directly as [pip-tools](https://github.com/jazzband/pip-tools) is doing.
    - https://github.com/sarugaku/resolvelib
    - others ?
4. Reformat the results in the `pyodide-lock.json` format (by taking information either from the original `pyodide-lock.json`)

Alternative proposals in https://github.com/pyodide/pyodide-lock/issues/8 https://github.com/pyodide/pyodide-lock/issues/9

**Comment by rth** [comment]
> I'd prefer the second option, to create a separate lockfile from the original pyodide-lock file, [...], and I suspect that this would cause version conflicts between duplicate packages.

@ryanking13  Not necessarily, I think, if you have some priority between channels ([--extra-index-url in pip](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-extra-index-url) pip option for instance) it should pick the version in the highest priority channel .

**Comment by ryanking13** [comment]
> Expose a Simple JSON API for the Pyodide distribution (so that Python package managers could understand it). This should already be done, but I can't find where (cc @ryanking13 )

Right now this is done in [pyodide-build/create_pypa_index.py](https://github.com/pyodide/pyodide/blob/main/pyodide-build/pyodide_build/create_pypa_index.py), so it can be used in `pip` after calling `pyodide venv`. By the way, it uses Simple HTML API but I think we better change it to Simple JSON API.

**Comment by rth** [comment]
> By the way, it uses Simple HTML API but I think we better change it to Simple JSON API.

Thanks! Absolutely. We cannot deploy HTML with the current CDN.

**Comment by ryanking13** [comment]
Among RFC 1, 2, and 3, IMO this is the best option for the long-term. I mean, in the future, we hope people can upload the Emscripten wheels to PyPI or to other package indices, and then we can create a lock file from it.

Then people who are okay with resolving dependencies dynamically can use Simple API (probably supported by micropip) to download packages from package indices, while people who make an application with fixed dependencies and don't want dependency resolution can create a lockfile.

**Comment by rth** [comment]
A Simple JSON API for this experimentation can be found in https://github.com/pyodide/pyodide/issues/3979

**Comment by rth** [comment]
@juntyr This could also be a useful direction to explore.

The main part is, take an example list of requirements, some standard Python packaging tool (TBD), specify that the supported wheels are only of the form `*cp311-emscripten_3_1_32_wasm32.whl` or `*py3-none-any.whl`, point it to the Pyodide's Simple JSON API (above) + PyPI and  ask it to generate a lock file (see the issue description above). If you have any questions let me know.

For instance, https://github.com/jazzband/pip-tools/issues/585 was essentially this feature request, but they decided it's out of scope for pip-compile.

Maybe @pradyunsg would also have some general advice or comments on this topic.

**Comment by webknjaz** [comment]
> Right now this is done in [pyodide-build/create_pypa_index.py](https://github.com/pyodide/pyodide/blob/main/pyodide-build/pyodide_build/create_pypa_index.py), so it can be used in `pip` after calling `pyodide venv`. By the way, it uses Simple HTML API but I think we better change it to Simple JSON API.

_(with my PyPA hat on)_
@ryanking13 could you rename PyPA (**Py**thon **Pa**ckaging **A**uthority) to PyPI (**Py**thon **P**ackages **I**ndex) in that script? It's kinda confusing to see.

> For instance, [jazzband/pip-tools#585](https://github.com/jazzband/pip-tools/issues/585) was essentially this feature request, but they decided it's out of scope for pip-compile.

That's not entirely accurate, but is a rather complicated topic. That issue was closed automatically because of a linked documentation PR, while there are other issues centered around the same discussion, including mine. Here's a few comments I made there that might be of interest when we're talking about "lockfiles": https://github.com/jazzband/pip-tools/issues/826#issuecomment-1324127670 / https://github.com/jazzband/pip-tools/issues/1326#issuecomment-1834517252.

Though, is it just a single platform in case of pyodide? FWIW it might be a good idea to participate in reviving PEP 665.

**Comment by ryanking13** [comment]
> @ryanking13 could you rename PyPA (Python Packaging Authority) to PyPI (Python Packages Index) in that script? It's kinda confusing to see.

Sure. Thanks for pointing that out.

**Comment by bollwyvl** [comment]
It seems like `uv` is now sufficient to do any "simple" out-of-band solves (e.g. don't involve `pyodide-build`) on most `uv`-supported platforms (e.g. not `pyodide` itself).

I've had some luck with this workflow
- build user-provided dependencies/excludes (including local wheels) into
  - `requirements.in`
  - `excludes.txt`
- build `pyodide-lock.json` into `constraints.txt` (maybe with some overrides)
- use `uv pip compile`to build build a PEP-571 `pylock.toml`
- fetch all remote wheels not present in the original
- run `PyodideLockSpec.add_wheels_to_spec`
- apply light patching after the fact, e.g.
  - fixing PEP-420 import name discovery
  - patching out known-bad `depends`

Would there be interest in PR proposing this, with a `pyodide-lock[uv]`?

As for running the above in `pyodide`: getting some subset of `uv` into a runtime `pyodide` itself would be nice, but a far bigger barge to pull.
